### ISSUE TITLE
Relational recurrent neural networks

### ISSUE THREAD

**Issue by flrngel** [flrngel/understanding-ai#17]
https://arxiv.org/abs/1806.01822
Paper from Deepmind

# Abstract
- paper's model built with intuition
- so it is unclear why this works well
- shows state-of-the-art on the WikiText103, Project Gutenberg, and GigaWord datasets

# 1. Introduction
- toy tested to stress current models; ***LSTMs***, ***DNC*** (from [differentiable neural computers](https://deepmind.com/blog/differentiable-neural-computers/))
- **Relational Memory Core**(**RMC**) uses
  - multi-head dot product attention (see [Transformer](https://arxiv.org/abs/1706.03762))
  - memory interactions

# 2. Relational reasoning
- this is process of understanding the ways in which entities are connected
- author claims that current models can be cast as relational reasoning
  - convolutional can be said to compute a relation of pixels (entities)
  - ***node*** and ***edge*** in message passing neural networks are entities
    - learnable nodes-edge connection can be relation
  - see also [[13](https://arxiv.org/abs/1609.02907), [14](https://arxiv.org/abs/1710.10903), [17](https://arxiv.org/abs/1711.07971)]
- but in current model, it is unclear whether inductive biases makes limitation
  - memory augmented networks 

# 3. Model
- paper's model is assembled with
  1. LSTMs
  2. memory augmented
  3. non-local net
- paper's model applied attention between memories at a single time step
  - this differs from current models that uses attention across all previous representations

## 3.1. Allowing memories to interact using MHDPA
- should note that ***MHDPA*** uses ***KQV*** linears
![image](https://user-images.githubusercontent.com/2807595/41603367-cacda4e8-7417-11e8-8c15-b798106719d9.png)
![image](https://user-images.githubusercontent.com/2807595/41603048-2292f418-7417-11e8-8b40-6553195b9feb.png)
equals to ![image](https://user-images.githubusercontent.com/2807595/41603089-38de6bb2-7417-11e8-9e79-99918401edee.png) which means proposed update to **M**(memory)

## 3.2. Encoding new memories
![image](https://user-images.githubusercontent.com/2807595/41603487-06f369da-7418-11e8-9b29-dac6e4a65bd3.png)

## 3.3. Introducing recurrence and embedding into an LSTM
see also [Official tensorflow implemenation](https://github.com/deepmind/sonnet/blob/master/sonnet/python/modules/relational_memory.py)
![image](https://user-images.githubusercontent.com/2807595/41603579-3c7df502-7418-11e8-9940-ba95317fd56a.png)

![image](https://user-images.githubusercontent.com/2807595/41603592-4326e3fa-7418-11e8-9e06-571d3d60acbb.png) means row/memory-wise MLP with layer normalization

# 4. Experiments
- they tested
  - **4.1. Illustrative supervised tasks**
    - N-th farthest
    - [Program Evaluation](https://arxiv.org/pdf/1806.01822)
  - **4.2. Reinforcement learning**
    - Mini Pacman with viewport
  - **4.3. Language Modeling**
    - WikiText103
    - Project Gutenberg
    - GigaWord datasets

![image](https://user-images.githubusercontent.com/2807595/41603845-c6dac0ae-7418-11e8-88db-0d6e1de7bf85.png)

# 5. Results
**Figure3** is from **n-th farthest** task
![image](https://user-images.githubusercontent.com/2807595/41603866-d300b776-7418-11e8-8055-ce54aebb9185.png)
![image](https://user-images.githubusercontent.com/2807595/41603914-e5880e58-7418-11e8-87e4-2c75d76b64a3.png)

# 6. Discussion
- author proposed inutions for the mechanisms for complex relational reasoning
- author considers their result primarily as "evidence of improved function"

# Appendix
![image](https://user-images.githubusercontent.com/2807595/41604447-3db45464-741a-11e8-9f90-5c681bdd50f0.png)

![image](https://user-images.githubusercontent.com/2807595/41604557-7ff805dc-741a-11e8-8759-b0ebd0fcc2c4.png)
